File: javascript/02_AboutArrays1/arrayDistanceAndPosition_02.js


// 2. ฟังก์ชันหาระยะห่างและตำแหน่งในอาเรย์ - ใช้วิเคราะห์ตำแหน่งและระยะห่างของค่าที่ซ้ำกันในอาเรย์
function arrayDistanceAndPosition(arr) {
    // ตรวจสอบว่าอาเรย์ว่างหรือเป็น null หรือไม่
    if (!arr || arr.length === 0) return null;
    
    // สร้างออบเจ็กต์เก็บผลลัพธ์
    const result = {
        minDistance: Infinity,           // ระยะห่างน้อยที่สุดระหว่างค่าที่ซ้ำกัน
        maxDistance: -Infinity,          // ระยะห่างมากที่สุดระหว่างค่าที่ซ้ำกัน
        positions: {}                    // เก็บตำแหน่งของแต่ละค่า
    };
    
    // วนลูปเก็บตำแหน่งของแต่ละค่าในอาเรย์
    arr.forEach((value, index) => {
        if (!result.positions[value]) {
            result.positions[value] = [];
        }
        result.positions[value].push(index);
    });
    
    // คำนวณระยะห่างระหว่างค่าที่ซ้ำกัน
    Object.values(result.positions).forEach(positions => {
        if (positions.length > 1) {
            for (let i = 1; i < positions.length; i++) {
                const distance = positions[i] - positions[i-1];
                result.minDistance = Math.min(result.minDistance, distance);
                result.maxDistance = Math.max(result.maxDistance, distance);
            }
        }
    });
    
    return result;
}

console.log(arrayDistanceAndPosition([1, 2, 1, 3, 2, 1]));

File: java/02_AboutArrays1/arrayDistanceAndPosition_02.java
import java.util.*;

public class arrayDistanceAndPosition_02 {
    public int minDistance;
    public int maxDistance;
    public Map<Integer, List<Integer>> positions;

    @Override
    public String toString() {
        return "ArrayAnalysis{" +
               "minDistance=" + minDistance +
               ", maxDistance=" + maxDistance +
               ", positions=" + positions +
               '}';
    }

    public static arrayDistanceAndPosition_02 arrayDistanceAndPosition(int[] array) {
        arrayDistanceAndPosition_02 analysis = new arrayDistanceAndPosition_02();
        analysis.positions = new HashMap<>();

        for (int i = 0; i < array.length; i++) {
            analysis.positions.computeIfAbsent(array[i], k -> new ArrayList<>()).add(i);
        }

        analysis.minDistance = Integer.MAX_VALUE;
        analysis.maxDistance = Integer.MIN_VALUE;

        for (List<Integer> posList : analysis.positions.values()) {
            if (posList.size() > 1) {
                for (int i = 1; i < posList.size(); i++) {
                    int distance = posList.get(i) - posList.get(i - 1);
                    analysis.minDistance = Math.min(analysis.minDistance, distance);
                    analysis.maxDistance = Math.max(analysis.maxDistance, distance);
                }
            }
        }

        if (analysis.minDistance == Integer.MAX_VALUE) {
            analysis.minDistance = 0; // No repeated elements
        }

        return analysis;
    }

    public static void main(String[] args) {
        // ทดสอบการวิเคราะห์อาเรย์
        System.out.println(arrayDistanceAndPosition(new int[]{1, 2, 1, 3, 2, 1}));
    }
}